Move the creation of Alpide chip in ITSMFT common part

mario6829 opened the pull request:
The creation of the Alpide chip was moved in the ITSMFT common area (the geometrical shape and daughters are the same as before). The V3Layer code was modified accordingly. In the shadow some small overlaps (due to the gaps between chips in IB modules) were cured.

mario6829 commented:
@iouribelikov @shahor02 Hi Yura and Ruben, I moved the code for the Alpide chip in the common ITSMFT directory. Please let me know if it is ok or something else is needed. Cheers Mario

shahor02 commented on Detectors/ITSMFT/ITS/simulation/src/V3Layer.cxx:
```diff
@@ -524,11 +526,23 @@ TGeoVolume *V3Layer::createModuleInnerB(Double_t xmod, Double_t ymod, Double_t z
+
   zchip = zmod / sIBChipsPerRow;
-  TGeoVolume *chipVol = createChip(xmod, ymod, zchip);
+  TGeoVolume *chipVol = AlpideChip::createChip(xmod, ymod, zchip,
```
Aren't xmod and zchip constants of the Alpide? In principle, the mSensorThickness also, but probably we can have it settable in case we will use different Miko's response objects for chips of different total thickness

mario6829 replied:
@shahor02 I took out the X and Z parameters and made those dimensions fixed in AlpideChip.h, please check if this is ok. I left the sensor thickness as a parameter (at least, for the time being) because you said we may try to use different sensor thicknesses for IB and OB.

shahor02 commented on Detectors/ITSMFT/common/simulation/include/ITSMFTSimulation/AlpideChip.h:
```diff
@@ -0,0 +1,60 @@
+
+ private:
+  static const Double_t sMetalLayerThick;    ///< Metal layer thickness
+
```
I would define this as ``static constexpr ...`` together with chip dX and dZ

mario6829 replied:
@shahor02 Done.

shahor02 replied:
@mario6829 thanks! For me this PR is fine.

mario6829 commented:
@shahor02 @iouribelikov Hello, I commited the modifications suggested by Ruben. @bovulpes if you want to change your implementation, please note that I changed the input parameter list as suggested by Ruben. Cheers Mario

bovulpes commented:
I am on the way of doing the changes, thanks.
Cheers,
Bogdan.

iouribelikov commented:
@mario6829 Hi Mario.  The code compiles and runs fine.

But, could you please check the positioning of the MetalStack ?  It should be sitting on the side (of the chip) that is oriented towards the FPC.    I am afraid I find it on the opposite side...

The same concerns the sensitive part :  It should be facing the FPC.

Thanks a lot and cheers,
Yura.

mario6829 commented:
@iouribelikov Hi Yura, yes you are right, I moved them in the right position, i.e. facing the FPC. Cheers
Mario

iouribelikov commented:
@mario6829  Thank you, Mario.  The ordering has definitely changed.
Now, it looks like this :  FPC -> Substrate -> MetalStack -> Sensitive.
Whereas it should be:  FPC -> MetalStack -> Sensitive -> Substrate.

Something is still missing. Could you have a deeper look, please ?

Thank you very much !
Cheers,
Yura.

alibuild commented:
Error while checking build/o2/macos for b24f82fe4ddd73da5712f6bdcffb2c2a7a86db14:
```

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/662/b24f82fe4ddd73da5712f6bdcffb2c2a7a86db14/fullLog.txt).

mario6829 commented:
@iouribelikov Hi Yura, from where do you find the aforementioned sequence of volumes ? If I draw the ITSUmodule volume with Root I found the correct sequence, and also if I inspect the position matrices of its daughter volumes I see the correct ordering. Indeed in AlpideChip::createChip the MetalStack is put at highest Y, then below it there is the ITSUSensor: so from above the order is Metal - Sensor - rest of the chip. The ITSUchip is placed without rotation in V3Layer::createModuleInnerB, so starting from below you have: Chip (Silicon - Sensor - Metal) - Glue - FPC Anode - Kapton - FPC Ground, which should be the correct order. Where do you find it different ?
Cheers
Mario

iouribelikov commented:
@mario6829  I am sorry.  I was doing the usual thing: drawing the geometry with "right click" in TBrowser.   But, I mis-counted the volumes, taking the glue layer for the substrate :-(
  
Everything is perfectly OK now.

Many thanks for your checks !
Cheers,
Yura.

bovulpes commented:
Hello @mario6829 ,

I have in my repository a branch "ms-dev" starting from your "dev" branch and with the MFT side of the modifications. Please use it for the MFT contribution to this PR.
Thanks.
cheers,
Bogdan.

mario6829 commented:
@bovulpes @iouribelikov @shahor02  Hi all, I managed to merge Bogdan's mods, could you please check if everything is ok ?
Thanks Cheers
Mario

alibuild commented:
Error while checking build/o2/macos for c035d0cbafb29cb9772d555137c08626f4f0e5f6:
```
sw/BUILD/FairRoot-latest/log
[ 48%] Linking CXX shared library ../lib/libFairMQ.dylib
[ 48%] Building CXX object base/CMakeFiles/Base.dir/event/FairMCEventHeader.cxx.o
ld: file not found: libboost_thread.dylib for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libFairMQ.dylib] Error 1
make[1]: *** [fairmq/CMakeFiles/FairMQ.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/662/c035d0cbafb29cb9772d555137c08626f4f0e5f6/fullLog.txt).

iouribelikov commented:
@mario6829 @bovulpes For me, everything works fine.
@sawenzel @shahor02  Shall we merge this PR ?

shahor02 commented on Detectors/ITSMFT/MFT/base/src/Geometry.cxx:
```diff
@@ -42,7 +42,8 @@ const Double_t Geometry::sSensorInterspace = 0.01; //[cm]  Offset between two ad
+const Double_t Geometry::sChipThickness = 50.e-4; // 50 microns
 
 const Double_t Geometry::sHeightActive = 1.3;
 const Double_t Geometry::sHeightReadout = 0.2;
```
These sHeightActive and sHeightReadout are not needed anymore, as far as I can see

shahor02 replied:
These sHeightActive and sHeightReadout also are not used

shahor02 commented on Detectors/ITSMFT/MFT/base/include/MFTBase/Geometry.h:
```diff
@@ -34,7 +36,8 @@ class Geometry : public TNamed {
+  static const Double_t sChipThickness;       ///< \brief CMOS chip thickness
   static const Double_t sXPixelPitch;         ///< \brief Pixel pitch along X
   static const Double_t sYPixelPitch;         ///< \brief Pixel pitch along Y
   static const Int_t sNPixelX = 1024;         ///< \brief Number of Pixel along X
```
These are also redundant

shahor02 commented on Detectors/ITSMFT/common/simulation/include/ITSMFTSimulation/AlpideChip.h:
```diff
@@ -0,0 +1,62 @@
+ private:
+  static constexpr Double_t sChipXWidth      =  1.5;   ///< Chip X width (cm)
+  static constexpr Double_t sChipZLength     =  3.0;   ///< Chip Z length (cm)
+  static constexpr Double_t sMetalLayerThick = 15.0*1.0E-4;  ///< Metal layer thickness (um)
```
Can we keep all Alpide dimensions constants in one single class? We have https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/ITSMFT/common/base/include/ITSMFTBase/SegmentationAlpide.h class which has almost everything, particularly SegmentationAlpide::SensorSizeCols and SensorSizeRows give public constexpr size along the columns and rows, i.e. sChipZLength and sChipXWidth

shahor02 commented:
@mario6829 , @bovulpes : I've cloned this branch to my https://github.com/shahor02/AliceO2/commits/dev_mario_alpide with the patch to use consistently Alpide constants defined in the SegmentationAlpide, also eliminated redundant (in few cases wrong) redefinitions in the MFT::Geometry.
Could you please check the only commit there:  https://github.com/shahor02/AliceO2/commit/b7f0614bcc3ebd8a526a59c3c5ffd69267820709 and squash to this PR if OK.

mario6829 commented:
@shahor02 @bovulpes @iouribelikov  Hi Ruben, I added your commit, please let me know if it's ok.
Cheers Mario

shahor02 commented:
@mario6829 thanks, for this PR is ok.

bovulpes commented:
Hello,
Checked the latest mario6829/dev and it is ok.
Thanks.

matthiasrichter commented on Detectors/ITSMFT/ITS/simulation/src/V3Layer.cxx:
```diff
@@ -524,11 +526,21 @@ TGeoVolume *V3Layer::createModuleInnerB(Double_t xmod, Double_t ymod, Double_t z
   Double_t xpos, ypos, zpos;
-  char volumeName[30];
+  Bool_t dummyChip;
+  char chipName[30], sensName[30], volumeName[30];
```
I suggest using `std::string` here, then it's safe even if the return of `GeometryTGeo::getITSChipPattern()` changes

matthiasrichter commented on Detectors/ITSMFT/ITS/simulation/src/V3Layer.cxx:
```diff
@@ -524,11 +526,21 @@ TGeoVolume *V3Layer::createModuleInnerB(Double_t xmod, Double_t ymod, Double_t z
-  zchip = zmod / sIBChipsPerRow;
-  TGeoVolume *chipVol = createChip(xmod, ymod, zchip);
+  snprintf(chipName, 30, "%s%d", GeometryTGeo::getITSChipPattern(), mLayerNumber);
+  snprintf(sensName, 30, "%s%d", GeometryTGeo::getITSSensorPattern(), mLayerNumber);
```
better use `std::string` operations, or make a constant variable for the `char[]` size, use it in the definitions and here. In principle, only max 29 characters can be copied here and a terminating `0` must be set at the end. No worries with `std::string` though

matthiasrichter commented on Detectors/ITSMFT/ITS/simulation/src/V3Layer.cxx:
```diff
@@ -524,11 +526,21 @@ TGeoVolume *V3Layer::createModuleInnerB(Double_t xmod, Double_t ymod, Double_t z
+  snprintf(sensName, 30, "%s%d", GeometryTGeo::getITSSensorPattern(), mLayerNumber);
+
+  TGeoVolume *chipVol = AlpideChip::createChip(ymod, mSensorThickness/2,
+					       chipName, sensName, dummyChip);
```
minor formatting issue, please replace tabs with space

matthiasrichter commented on Detectors/ITSMFT/ITS/simulation/src/V3Layer.cxx:
```diff
@@ -570,6 +582,7 @@ TGeoVolume *V3Layer::createModuleInnerB(Double_t xmod, Double_t ymod, Double_t z
   // Build up the module
   xpos = -xtot + ((TGeoBBox*)chipVol->GetShape())->GetDX() + sIBFPCWiderXNeg;
   ypos = -ytot + ymod; // = 0 if not kIBModel4
+  zchip = ((TGeoBBox*)chipVol->GetShape())->GetDZ();
```
probably use a pointer cast here

matthiasrichter commented on Detectors/ITSMFT/ITS/simulation/src/V3Layer.cxx:
```diff
@@ -3261,7 +3204,7 @@ TGeoVolume *V3Layer::createModuleOuterB(const TGeoManager *mgr)
 
 
-  char volname[30];
+  char chipName[30], sensName[30], volName[30];
```
`std::string` also here or at least constant variable for array size, used here and in `snprintf`

matthiasrichter commented on Detectors/ITSMFT/common/simulation/include/ITSMFTSimulation/AlpideChip.h:
```diff
@@ -0,0 +1,59 @@
+
+  static constexpr Double_t sMetalLayerThick = 15.0*1.0E-4;  ///< Metal layer thickness (um)
+
+  ClassDef(AlpideChip, 0) // AlpideChip geometry
```
use `ClassDefNV`

matthiasrichter commented on Detectors/ITSMFT/common/simulation/src/AlpideChip.cxx:
```diff
@@ -0,0 +1,121 @@
+				   const char* chipName,
+				   const char* sensName,
+				   const Bool_t dummy,
+				   const TGeoManager *mgr){
```
replace tabs with spaces

mario6829 commented:
Hello @matthiasrichter  many thanks for your suggestions. I'll fix the blank spaces and commit the changes. For what concerns the proper use of strings, do you think it would be acceptable to postpone the fix to a subsequent PR ? I ask this because I have already other pending fixes on good coding practices Sandro recommended me to implement, so I would prefer to gather all those code fixes on a dedicated PR.
Cheers   Mario

sawenzel commented:
@mario6829 : I am ok with a later "proper" fix. Let me know when you have addressed the smaller changes.

mario6829 commented:
@sawenzel @matthiasrichter Hello, I replaced Tabs with Blanks wherever I found them in the code (there were quite a lot - I configured my editor to emit blanks when tabbing) both for V3Layer and AlpideChip classes. I also changed ClassDef into ClassDefNV. Let me know if this is fine with you.
I opened a Jira ticket setting you as watchers to keep track of the improvements regarding the good C++ coding I have to fix.
Cheers
Mario
